openforcefield/topology/topology.py

```diff
@@ -1547,20 +1547,86 @@ def from_openmm(cls, openmm_topology, unique_molecules=None):
         # TODO: How can we preserve metadata from the openMM topology when creating the OFF topology?
         return topology
 
-    # TODO: Jeff prepended an underscore on this before 0.2.0 release to remove it from the API.
-    #       Given the recent (2019_04) discussions about potential loss of parameters during conversion, we should
-    #       revisit this function to determine what sorts of guarantees we can put on system correctness before
-    #       we expose it.
-    def _to_openmm(self):
+    def to_openmm(self):
         """
         Create an OpenMM Topology object.
 
+        The OpenMM ``Topology`` object will have one residue per topology
+        molecule. Currently, the number of chains depends on how many copies
+        of the same molecule are in the ``Topology``. Molecules with more
+        than 5 copies are all assigned to a single chain, otherwise one
+        chain is created for each molecule. This behavior may change in
+        the future.
+
         Parameters
         ----------
         openmm_topology : simtk.openmm.app.Topology
             An OpenMM Topology object
         """
-        raise NotImplementedError
+        from simtk.openmm.app import Topology as OMMTopology
+        from simtk.openmm.app import Single, Double, Triple, Aromatic
+        from simtk.openmm.app.element import Element as OMMElement
+
+        omm_topology = OMMTopology()
+
+        # Keep track of which chains and residues have been added.
+        mol_to_chains = {}
+        mol_to_residues = {}
+
+        # Go through atoms in OpenFF to preserve the order.
+        omm_atoms = []
+        # We need to iterate over the topology molecules if we want to
+        # keep track of chains/residues as Atom.topology_molecule is
+        # instantiated every time and can't be used as a key.
+        for topology_molecule in self.topology_molecules:
+            for atom in topology_molecule.atoms:
+                reference_molecule = topology_molecule.reference_molecule
+                n_molecules = len(self._reference_molecule_to_topology_molecules[reference_molecule])
```

Comment:
I'm glad you got a chance to see the madness that is the TopologyMolecule class :-P I wished for memory efficiency, but at what cost?